### PR TITLE
User/aeloros/get args

### PR DESCRIPTION
**Pull request by aeloros**
Implement the first set of functionality for the AppLifecycle and ActivationRegistrationManager objects.  
Support in this change is scoped to implementation of the LaunchActivatedEventArgs and ProtocolActivatedEventArgs scenarios for unpackaged Win32.
Sample APIs were removed from the tree.

Testing was done by adding basic launch tests, and testing protocol scenarios with a basic sample app.  A later PR will contain protocol tests, and a move to the new test framework.

### CODE REVIEW

File: dev/AppLifecycle/AppLifecycle.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "LaunchActivatedEventArgs.h"
+#include "ProtocolActivatedEventArgs.h"
+#include "Shared.h"
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    std::tuple<std::wstring, std::wstring> AppLifecycle::ParseCommandLine(std::wstring commandLine)
+    {
+        auto argsStart = commandLine.rfind(L"----") + 4;
```
Comment (jonwis) [inline dev/AppLifecycle/AppLifecycle.cpp:16]: `rfind` and `find_first_of` returns `npos` if the thing isn't there.  Consider adding checks for npos before doing additional math.

File: test/AppLifecycle/ApiTests.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            Assert::IsTrue(winrt::Microsoft::ProjectReunion::AppLifecycle::GetActivatedEventArgs() != nullptr);
+        }
+
+        TEST_METHOD(GetActivatedEventArgsForLaunch)
+        {
+            auto args = winrt::Microsoft::ProjectReunion::AppLifecycle::GetActivatedEventArgs();
+            Assert::IsTrue(args != nullptr);
+            Assert::IsTrue(args.Kind() == ActivationKind::Launch);
+
+            auto launchArgs = args.as<LaunchActivatedEventArgs>();
```
Comment (jonwis) [inline test/AppLifecycle/ApiTests.cpp:33]: Is this a Microsoft type or a Windows type here? The usings above confuse me.

**Review by jonwis — CHANGES_REQUESTED** [review]

**Review by dabhattimsft — APPROVED** [review]

File: dev/AppLifecycle/ActivatedEventArgsBase.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            return m_previousState;
+        }
+
+        SplashScreen SplashScreen()
+        {
+            return m_splashScreen;
+        }
+
+    protected:
+        ActivatedEventArgsBase()
```
[the remark this answers is not in the document]
Comment (aeloros) [reply dev/AppLifecycle/ActivatedEventArgsBase.h]: Updated

File: test/CppTest_UWP/pch.h [hunk 4]
```diff
@@ -19,6 +19,7 @@
 #include <winrt/Microsoft.VisualStudio.TestPlatform.TestExecutor.WinRTCore.h>
 #include <CppUnitTest.h>
 
+#include <winrt/Windows.ApplicationModel.Activation.h>
```
[the remark this answers is not in the document]
Comment (aeloros) [reply test/CppTest_UWP/pch.h:22]: Windows.

File: test/AppLifecycle/ApiTests.cpp (hunk 2 above)
Comment (aeloros) [reply test/AppLifecycle/ApiTests.cpp:33]: The interfaces are defined by the platform.  The concrete types are implemented by this library.  So the interfaces are Windows types, and the concreate types are Microsoft.  Does that help clarify?

File: dev/AppLifecycle/ActivatedEventArgsBase.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        {
+        }
+
+        ActivationKind m_kind;
+        ApplicationExecutionState m_previousState;
```
Comment (EHO-makai) [inline dev/AppLifecycle/ActivatedEventArgsBase.h:33]: does this have a reasonable initial value?

File: dev/AppLifecycle/ActivationRegistrationManager.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+#include "ProtocolActivatedEventArgs.h"
+#include "Shared.h"
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    void ActivationRegistrationManager::RegisterForFileTypeActivation(hstring const& groupName,
+        hstring const& logo, array_view<hstring const> supportedFileTypes,
+        array_view<hstring const> supportedVerbs)
+    {
+        throw hresult_not_implemented();
```
Comment (EHO-makai) [inline dev/AppLifecycle/ActivationRegistrationManager.cpp:17]: I defer to Jon on whether we are allowing "not impls" into main.

File: dev/AppLifecycle/ActivationRegistrationManager.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    void ActivationRegistrationManager::RegisterForProtocolActivation(hstring const& scheme,
+        hstring const& displayName)
+    {
+        if (scheme.empty() || displayName.empty())
+        {
+            throw winrt::hresult_invalid_argument();
+        }
+
+        if (HasIdentity())
```
Comment (EHO-makai) [inline dev/AppLifecycle/ActivationRegistrationManager.cpp:28]: Our only scenario for release 0.1 is packaged win32, which has identity.  Will this cause it to not work?  Or protocol activation is not supported currently for packaged apps?

File: dev/AppLifecycle/AppLifecycle.cpp [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+        // We explicitly use find_first_of here, so that the resulting data may contain : as a valid character.
+        auto argsEnd = commandLine.find_first_of(L":", argsStart);
+        auto argsLength = argsEnd - argsStart;
```
Comment (EHO-makai) [inline dev/AppLifecycle/AppLifecycle.cpp:20]: is there reasonable way to do what intsafe.h does here?

File: dev/AppLifecycle/AppLifecycle.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        }
+        else
+        {
+            // Generate IActivatedEventArgs for non-Packaged applications.
+            std::wstring contractId;
+            std::wstring contractData;
+            auto commandLine = std::wstring(GetCommandLine());
+            std::tie(contractId, contractData) = ParseCommandLine(commandLine);
+
+            if (contractId == c_protocolArgumentString)
```
Comment (EHO-makai) [inline dev/AppLifecycle/AppLifecycle.cpp:40]: more defensive? 
if (protocol)
{
do protocol
}
else if (launch)
{
do launch
}
else
{
throw unexpected
}

?

File: dev/AppLifecycle/ProtocolActivatedEventArgs.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    using namespace winrt::Windows::Foundation::Collections;
+    using namespace winrt::Windows::ApplicationModel::Activation;
+
+    class ProtocolActivatedEventArgs : public winrt::implements<ProtocolActivatedEventArgs, 
+        ActivatedEventArgsBase, IProtocolActivatedEventArgs>
+    {
+    public:
+        ProtocolActivatedEventArgs(const std::wstring& uri) : m_uri(winrt::Windows::Foundation::Uri(uri))
+        {
+            m_kind = ActivationKind::Protocol;
```
[the remark this answers is not in the document]
Comment (EHO-makai) [reply dev/AppLifecycle/ProtocolActivatedEventArgs.h:19]: I like that suggestion ...

File: dev/AppLifecycle/Shared.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include <pch.h>
+#include "Shared.h"
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    std::wstring GetFullIdentityString()
+    {
+        std::wstring identityString;
+        WCHAR idNameBuffer[PACKAGE_FULL_NAME_MAX_LENGTH];
```
Comment (EHO-makai) [inline dev/AppLifecycle/Shared.cpp:11]: do you need a +1 for a null term char?

File: dev/AppLifecycle/AppLifecycle.h [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    struct AppLifecycle
+    {
+        AppLifecycle() = default;
+
+        static Windows::ApplicationModel::Activation::IActivatedEventArgs GetActivatedEventArgs();
+
+    private:
+        static std::tuple<std::wstring, std::wstring> ParseCommandLine(std::wstring commandLine);
```
[the remark this answers is not in the document]
Comment (aeloros) [reply dev/AppLifecycle/AppLifecycle.h:16]: Without making this a global C function, I'm not sure of what you're suggesting.  A global C function has a broader scope, and this is just a portioned off piece of functionality of the object that doesn't require any instance state.  Are you suggesting some new technique I'm unaware of maybe?

File: dev/AppLifecycle/Shared.cpp [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    bool HasIdentity()
+    {
+        return !(GetFullIdentityString()).empty();
+    }
+
+    std::wstring GetModulePath()
+    {
+        std::wstring path(100, L'?');
```
Comment (EHO-makai) [inline dev/AppLifecycle/Shared.cpp:28]: why 100?

File: dev/AppLifecycle/Shared.cpp [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+        std::wstring defaultValue{ L"URL:" };
+        defaultValue += displayName;
+        THROW_IF_WIN32_ERROR(::RegSetValueEx(key.get(), nullptr, 0, REG_SZ,
+            reinterpret_cast<BYTE const*>(defaultValue.c_str()),
+            static_cast<uint32_t>((defaultValue.size() + 1) * sizeof(wchar_t))));
+
+        std::wstring urlProtocolValue{ L"" };
+        defaultValue += displayName;
+        THROW_IF_WIN32_ERROR(::RegSetValueEx(key.get(), L"URL Protocol", 0, REG_SZ,
```
Comment (EHO-makai) [inline dev/AppLifecycle/Shared.cpp:70]: did you want to define a constant for this somewhere?

File: dev/AppLifecycle/AppLifecycle.idl [hunk 15]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.ProjectReunion
```
Comment (EHO-makai) [inline dev/AppLifecycle/AppLifecycle.idl:4]: @jonwis , I was imagining that we would not have anything in "Microsoft.ProjectReuion" namespace.  Is this OK?

File: test/CppTest/CppTest.vcxproj [hunk 16]
```diff
@@ -111,6 +111,9 @@
   <ImportGroup Label="Shared">
     <Import Project="..\CppShared\CppShared.vcxitems" Label="Shared" />
   </ImportGroup>
+  <ImportGroup Label="AppLifecycle">
```
Comment (EHO-makai) [inline test/CppTest/CppTest.vcxproj:114]: you should have a packaged win32 (Centennial) test.

File: dev/AppLifecycle/AppLifecycle.cpp (hunk 1 above)
Comment (aeloros) [reply dev/AppLifecycle/AppLifecycle.cpp:16]: Updated on latest iteration.

File: dev/AppLifecycle/ActivationRegistrationManager.cpp (hunk 7 above)
Comment (aeloros) [reply dev/AppLifecycle/ActivationRegistrationManager.cpp:28]: Dynamic registration (@ runtime) of extension contracts will not be supported for packaged applications.  Instead they require entries to be placed in their manifest in order to register.

File: dev/AppLifecycle/AppLifecycle.cpp (hunk 8 above)
Comment (aeloros) [reply dev/AppLifecycle/AppLifecycle.cpp:29]: Elaborate please.

File: dev/AppLifecycle/AppLifecycle.cpp (hunk 9 above)
Comment (aeloros) [reply dev/AppLifecycle/AppLifecycle.cpp:54]: No.  Launch is our default, and has no heuristic other than "not others".

File: dev/AppLifecycle/AppLifecycle.cpp (hunk 8 above)
Comment (EHO-makai) [reply dev/AppLifecycle/AppLifecycle.cpp:29]: subtraction has the potential for overflow.  intsafe has helpers that return an error in this case,

File: dev/AppLifecycle/AppLifecycle.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (jonwis) [reply dev/AppLifecycle/AppLifecycle.h:16]: I'm saying that if it's a private static method on a class in a header, nobody but members of that class can use it anyway. So just don't put it in the header at all - put it at the top of applifecycle.cpp as a free function.  Delete it here, and at the top of applifecycle.cpp put `std::tuple<std::wstring, std::wstring> ParseCommandLine(std::wstring const& commandLine)`

File: dev/AppLifecycle/ActivationRegistrationManager.cpp (hunk 6 above)
Comment (jonwis) [reply dev/AppLifecycle/ActivationRegistrationManager.cpp:17]: IIRC there's parts of this overall effort that require additional work, right? Integration with Project Reunion toasting, for instance. I'd _prefer_ there's an implementation, or one coming soon.  We should avoid having a release with this API present but not implemented yet.

File: dev/AppLifecycle/AppLifecycle.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (aeloros) [reply dev/AppLifecycle/AppLifecycle.h]: Updated

File: dev/AppLifecycle/ActivatedEventArgsBase.h (hunk 5 above)
Comment (aeloros) [reply dev/AppLifecycle/ActivatedEventArgsBase.h:33]: Updated

File: dev/AppLifecycle/ProtocolActivatedEventArgs.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (aeloros) [reply dev/AppLifecycle/ProtocolActivatedEventArgs.h:19]: I do to.  I couldn't get it to work, and wonder if it's related to the winrt::implements<> class.  I'll fix this later after talking to Kenny.

File: dev/AppLifecycle/Shared.cpp (hunk 11 above)
Comment (aeloros) [reply dev/AppLifecycle/Shared.cpp]: Updated

File: dev/AppLifecycle/Shared.cpp (hunk 14 above)
Comment (aeloros) [reply dev/AppLifecycle/Shared.cpp:70]: I'll update later.

**Review by EHO-makai — APPROVED** [review]
There will be additional changes for the file activation at which time we may change the namespace,

File: dev/AppLifecycle/AppLifecycle.idl (hunk 15 above)
Comment (jonwis) [reply dev/AppLifecycle/AppLifecycle.idl:4]: For now it is, but we should probably move this to Ms.AppModel.Something eventually

**Review by jonwis — APPROVED** [review]

**Comment by jevansaks** [comment]
@aeloros please change the title of this PR to have a human-readable one-line-description of this PR (Use the "Edit" button on the top right of the PR page)

**Pull request merged**